### DISCUSSION TITLE
ngalert: Pagerduty Notifier Improvements Contribution Guidance

### DISCUSSION THREAD

**Discussion by bruno-dasilva** [grafana/grafana#50814] [Alerting]
Our company uses pagerduty current and we have some improvements we’d like to see with the Pagerduty notifier - improvements we could potentially contribute with some guidance. Namely these two:

1. Per-alert label overrides. Currently we’d have to create a metric ton of contact points in ngalert to match each unique Service x Severity x Component etc. In old alerting, we could override those params as labels in the triggered alert rules. https://github.com/grafana/grafana/discussions/47660
1. Passing json as “CustomDetails” to pagerduty instead of formatted text. Current implementation uses the default `__text_alert_list` template. Pagerduty’s events api can parse this if it’s json but not if its text. https://github.com/grafana/grafana/blob/v8.1.2/pkg/services/ngalert/notifier/channels/pagerduty.go#L65

Now 1. is pretty straightforward, so long as we’re cool to follow a similar implementation as in the old alerting implentation. Can someone greenlight this approach? https://github.com/grafana/grafana/blob/v8.1.2/pkg/services/alerting/notifiers/pagerduty.go#L138

For 2. it requires a tiny bit of extra design thought — I assume we don’t want to break the old behaviour and therefore want to keep text format as default. How best with the current templating setup can we achieve this? Should we add a flag to the UI and do json marshalling in code, or should we provide another template config block and let users specify a text/template override in-ui? Or maybe some other approach entirely?

If both of these are implemented, the Pagerduty integration becomes 10x more useful. With json fields AND overrides, we can send every alert to a single, global pagerduty event endpoint and then use in-pagerduty rules to add conditions and route events to the right destinations. https://support.pagerduty.com/docs/event-management#event-orchestration-and-rulesets

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
